service/s3/s3crypto/authentication_mode.go

```diff
@@ -0,0 +1,50 @@
+package s3crypto
+
+import (
+	"io"
+)
+
+// AuthenticationMode will use AES GCM for the main cipher.
+type AuthenticationMode struct {
+	// Cipher will hold either a master symmetric key or a KMS client
+	// for encrypting the Envelope key
+	BaseKeyProvider
+
+	// This is the CEK algorithm used
+	CipherData
+}
+
+// Authentication returns a new encryption only mode structure with a specific cipher
+// for the master key
+func Authentication(kp KeyProvider) CryptoMode {
+	return &AuthenticationMode{
+		BaseKeyProvider: BaseKeyProvider{kp},
+		CipherData:      CipherData{"AES/GCM/NoPadding", "128"},
+	}
+}
+
+// EncryptContents will generate a random key and iv and encrypt the data using cbc
+func (mode *AuthenticationMode) EncryptContents(dst io.Writer, src io.Reader) error {
+	// Sets the key and iv to a randomly generated key and iv
```

Comment:
Updating `EncryptContents ` to follow the streaming pattern would simplify this function a little bit removing the need for the `io.Copy`. The `io.Copy` effectively would get moved into `newSHA256Writer` I think and,  updated to take a reader instead of being written to.